src/cascade/model/session.py

```diff
@@ -0,0 +1,552 @@
+from collections import Iterable
+from contextlib import contextmanager
+from math import nan
+from pathlib import Path
+from subprocess import run, PIPE
+
+import numpy as np
+import pandas as pd
+
+from cascade.core import getLoggers
+from cascade.dismod.constants import DensityEnum, RateEnum, INTEGRAND_TO_WEIGHT, IntegrandEnum, COMMAND_IO
+from cascade.dismod.db.wrapper import DismodFile, get_engine
+from cascade.dismod.serialize import default_integrand_names, make_log_table
+from cascade.model import Model
+from cascade.model.model_reader import (
+    read_var_table_as_id, read_vars, write_vars, read_prior_residuals, read_data_residuals, read_samples
+)
+from cascade.model.model_writer import ModelWriter
+
+CODELOG, MATHLOG = getLoggers(__name__)
+
+
+class Session:
+    def __init__(self, locations, parent_location, filename):
+        """
+        A session represents a connection with a Dismod-AT backend through
+        a single Dismod-AT db file, the sqlite file it uses for input and
+        output.
+
+        Args:
+            locations (pd.DataFrame): Initialize here because data refers to this.
+            parent_location (int): The session uses parent location to subset
+                                   data, but it isn't in the model.
+            filename (str|Path): Location of the Dismod db to overwrite.
+        """
+        self.dismod_file = DismodFile()
+        self._filename = Path(filename)
+        if self._filename.exists():
+            MATHLOG.info(f"{self._filename} exists so overwriting it.")
+            self._filename.unlink()
+        self.dismod_file.engine = get_engine(self._filename)
+        self.parent_location = parent_location
+
+        self._create_node_table(locations)
+        self._basic_db_setup()
```

Comment:
Even if _basic_db_setup doesn't do anything that _create_node_table depends on it is arresting to see the setup function called after the first function that touches the database.